### PR TITLE
Added a damage prevention config option for Charge Up.

### PR DESCRIPTION
**Pull request by Delfofthebla**
Added a damage prevention config option for Charge Up. This will allow users to configure just how much damage is prevented from the ability.

I personally felt that this skill was just far too powerful with the current setup. Ideally I'd set it anywhere from 0 to 50% damage reduction, so I felt that a config option would be best.

### CODE REVIEW

**Comment by Delfofthebla** [comment]
@TheTimeSweeper Lemme know what you think

**Comment by TheTimeSweeper** [comment]
Shoulda seen it before there was a limit on damage return :lul:
Have you tested this? The guy has no mobility, his defensive option needed to be good. In ror2, damage reduction even up to 50% lends you to being one-shot late game so I'm skeptical of how it feels over the consistency of invincibility.

**Comment by Delfofthebla** [comment]
I've tested that the code works, but I haven't done a full run with a lowered value yet. The last time I played Tesla I got an afterburner super early and was kinda bummed that I had literally no way to lose lol. Which is what inspired me to make these changes. 

I can do a few practice runs with the number reduced if you'd like but to be honest the entire reason I made a config option  was so that anyone using the mod could just kinda play it by ear. There's definitely other survivors with no mobility and they make do even without any substantial defense options. Quail, Hopoo, and just plain ol' movespeed can go a long way.

**Comment by Delfofthebla** [comment]
Ok. I've played a few games over the past few days with the config option. With numbers ranging from 0% damage block to 75% damage block

Ultimately, I agree with your statement that the Trooper has survivability problems. He is effectively a melee survivor and with no mobility or typical defensive options, More-over, his utility requires him to actually go out of his way to intentionally take damage if he wants to deal damage with it--requiring him to stay in melee for up to 6 seconds to "soak" damage, and a short animation afterwards to deal it.

By having the config option set to anything less than 100%, it feels inherently risky and unnecessary to do this, as dying is relatively easy and getting yourself out of danger after the explosion is not really possible until you are very late in the game. In the late game, a number around 50% felt fine. I used it less as a damage utility and more as a preparation button for incoming damage. The casting animation made the skill very risky to use however, and so I no longer spammed it as much as I did prior to my changes.

---

I see two problems with Charge Up! as it exists in its current form.
1) Tesla Trooper is too weak early and depends upon his invulnerability to guard him against his lack of movement speed and lack of damage. The current implementation is required for early game run consistency. Without 100% damage prevention, he is very likely to lose within stages 1-3, and his progress is very slow in these stages as well. He feels very similar to REX in that if you do not acquire certain key items early enough, your run will come to a very quick end.

2) Charge Up! is grossly overpowered in the mid to late game (stages 6+). At this stage of the game he no longer requires Charge Up for damage and can simply rely on proc chains, backup mags, Lysate Cell, or Ancient Scepter to cleanup large groups of enemies. However, it offers unparalleled survivability. Mercenary' invuln abuse is the closest I can think of in terms of power.

If you manage to acquire enough cooldown reduction (and/or extra charges) via Bandolier, Alien Head(s), Afterburner, or even the less desirable 'Purity' lunar item; you effectively make the trooper invulnerable. Permanent or near permanent invulnerability was a constant issue I ran into during loop 2/3.

---

The goal of this Pull Request was simply to give individual players the ability to customize the mod in a way that better suited the way they played the game--not to replace the existing skill behavior. I think my code accomplishes that, and does it in a way that people who want to keep the invulnerability are able to do so. It is the default behavior with this config and I think that's fine.

Honestly I think I'd rather see a rework to the skill behavior to alleviate the two problems I mentioned up above. Maybe having the skill grant movement speed, or deal the "absorbed damage" to him over time after the explosion occurs, or maybe just a different skill concept all together, I'm not sure. But that's a lot more work than just a single config option, so here I am.

**Comment by TheTimeSweeper** [comment]
Went ahead and merged it by accident (pulled it to a separate branch to check it out then did work on that branch without realizing lol so I had to just merge everything).
I'll be thinking it over more. My philosophy on configs, at least towards the beginning of the mod's lifespan is a little strict when it comes to balance stuff. I have some fun ones in there that do mess with balance, but as far as something that's a core balance decision like the utility I'd want to keep that under control until it's in a good spot, then maybe introduce customizability later.
As for the move itself, thanks so much for looking into it so deeply. I still want to stand by the invincibility, as it's just fun that way, but feeling too strong is a notion I've gotten from a few people now. I'll feel it out myself more, but I think I'll start with a nerf to downtime/uptime ratio. I imagine that's not gonna affect much early game, cause you're not completely taking advantage of the full 4 seconds anyways, and later game with refreshers and cooldown reduction it'll still be useful. If it's still too strong with items, I might make it somewhere like 80-90% there, so it's still mostly invincible, but there's at least the *opportunity* to get hit. 
In conclusion, idk have a lovely evening. This was quite helpful thanks thanks
